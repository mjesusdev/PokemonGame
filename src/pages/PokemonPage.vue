<template>
    <div class="row justify-content-center">
        <div class="col-lg-8">
            <h1 class="text-dark" v-if="!pokemon">Espere por favor...</h1>

            <div v-else>
                <img class="img-fluid" src="/img/pokemon-game.png" alt="Imagen Pokemon Game">
                <h1 class="text-info mt-4">Â¿QuiÃ©n es este pokÃ©mon?</h1>

                <PokemonPicture 
                    :pokemonId="pokemon.id" 
                    :showPokemon="showPokemon"
                />

                <PokemonOptions 
                    :pokemons="pokemonArr"
                    @selection-pokemon="checkAnswer"
                />

                <template v-if="showAnswer">
                    <h2 class="fw-bold fade-in mt-4 mb-4" :class="type">{{ message }}</h2>
                    <button class="btn btn-lg pb-3 btnNewGame" @click="newGame">
                        Nuevo Juego
                    </button>
                </template>
            </div>
        </div>
    </div>
</template>

<script>
    import PokemonOptions from '@/components/PokemonOptions'
    import PokemonPicture from '@/components/PokemonPicture'
    import getPokemonOptions from '@/helpers/getPokemonOptions'

    export default {
        components: { PokemonOptions, PokemonPicture  },
        data() {
            return {
                pokemonArr: [],
                pokemon: null,
                showPokemon: false,
                showAnswer: false
            }
        },
        methods: {
            async mixPokemonArray() {
                this.pokemonArr = await getPokemonOptions()

                const rndInt = Math.floor( Math.random() * 4 )
                this.pokemon = this.pokemonArr[ rndInt ]
            },
            checkAnswer( selectedId ) {
                this.showPokemon = true
                this.showAnswer  = true

                if ( selectedId === this.pokemon.id ) {
                    this.message = `ðŸŽ‰ Â¡Correcto! es ${ this.pokemon.name } ðŸŽ‰`
                    this.type = 'text-success'
                } else {
                    this.message = `ðŸ™ˆ Â¡Incorrecto! era ${ this.pokemon.name } ðŸ™ˆ`
                    this.type = 'text-danger'
                }
            },
            newGame() {
                this.showPokemon = false
                this.showAnswer  = false
                this.pokemonArr  = []
                this.pokemon     = null
                this.mixPokemonArray()    
            }
        },
        mounted() {
            this.mixPokemonArray()
        }
    }
</script>

<style scoped>
    .btnNewGame {
        background: #4baf8a !important;
        border-color: #4baf8a !important;
        color: #fff !important;
    }

    .btnNewGame:hover {
        background: #24c88c !important;
    }

    .text-info {
        color: #b8d3ff !important;
    }

    .text-success {
        color: #4baf8a !important;
    }

    .text-danger {
        color: #f44545 !important;
    }
</style>