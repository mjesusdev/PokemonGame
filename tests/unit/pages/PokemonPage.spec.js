import { shallowMount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })
    
    test('Should do match with Snapshot', () => {
        expect( wrapper.html() ).toMatchSnapshot()
    })
    
    test('Should call the function mixPokemonArray to mount', () => {
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        const wrapper = shallowMount( PokemonPage )
        expect( mixPokemonArraySpy ).toHaveBeenCalled();
    })

    test('Should do match with Snapshot when load the pokemons', () => {
        const wrapper = shallowMount( PokemonPage ,{ 
            data () {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('Should show the components of PokemonPicture and PokemonOptions', () => {
        const wrapper = shallowMount( PokemonPage ,{ 
            data () {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect( picture.exists() ).toBeTruthy()
        expect( options.exists() ).toBeTruthy()

        expect( picture.attributes('pokemonid') ).toBe('5')
        expect( options.attributes('pokemons') ).toBeTruthy()
    })

    test('test with checkAnswer', async() => {
        const wrapper = shallowMount( PokemonPage ,{ 
            data () {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(5)
    
        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe(true)
        expect( wrapper.find('h2').text() ).toBe(`ðŸŽ‰ Â¡Correcto! es ${ pokemons[0].name } ðŸŽ‰`)

        await wrapper.vm.checkAnswer(6)

        expect( wrapper.vm.message ).toBe(`ðŸ˜† Â¡Incorrecto! era ${ pokemons[0].name } ðŸ˜…`);
    })


})