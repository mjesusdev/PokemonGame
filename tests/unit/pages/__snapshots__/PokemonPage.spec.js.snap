// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Should do match with Snapshot 1`] = `
<div class="row justify-content-center">
  <div class="col-lg-8">
    <h1 class="text-dark">Espere por favor...</h1>
  </div>
</div>
`;

exports[`PokemonPage Component Should do match with Snapshot when load the pokemons 1`] = `
<div class="row justify-content-center">
  <div class="col-lg-8">
    <div><img class="img-fluid" src="/img/pokemon-game.png" alt="Imagen Pokemon Game">
      <h1 class="text-info mt-4">¿Quién es este pokémon?</h1>
      <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
      <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
      <!--v-if-->
    </div>
  </div>
</div>
`;
